File: core/encounters.py
import random
from typing import Dict, Iterable, List

from core.entities import Enemy


ENEMY_DEFINITIONS: Dict[str, dict] = {
    "shadow": {
        "name": "Shadow",
        "hp": 22,
        "atk": 2,
        "defense": 1,
        "cd": 2.5,
        "portrait_path": "assets/portraits/enemies/shadow.png",
        "xp_reward": 12,
        "munny_reward": 6,
        "drops": [
            {
                "material_id": "bright_shard",
                "chance": 0.5,
                "amount": 1,
            },
            {
                "material_id": "dark_shard",
                "chance": 0.4,
                "amount": 1,
            },
        ],
    },
    "soldier": {
        "name": "Soldier",
        "hp": 24,
        "atk": 2,
        "defense": 1,
        "cd": 2.0,
        "portrait_path": "assets/portraits/enemies/soldier.png",
        "xp_reward": 24,
        "munny_reward": 12,
        "drops": [
            {
                "item_id": "champion_belt",
                "chance": 0.25,
            },
            {
                "material_id": "mythril_fragment",
                "chance": 0.5,
                "amount": 1,
            },
        ],
    },
}


class EncounterPool:
    """Manage enemy templates and spawn new instances on demand."""

    def __init__(
        self,
        pools: Dict[str, Iterable[dict]],
        *,
        default_pool: str,
        enemy_definitions: Dict[str, dict] | None = None,
    ):
        if default_pool not in pools:
            raise KeyError(f"Unknown enemy pool '{default_pool}'")
        # Copy each template so edits here never touch caller data.
        self._pools: Dict[str, List[dict]] = {}
        for name, entries in pools.items():
            self._pools[name] = [dict(entry) for entry in entries]
        self._current_pool = default_pool
        self._rng = random.Random()
        self._enemy_definitions = {
            key: dict(value)
            for key, value in (enemy_definitions or ENEMY_DEFINITIONS).items()
        }

    @property
    def current_pool(self) -> str:
        return self._current_pool

    def set_pool(self, pool_name: str) -> None:
        if pool_name not in self._pools:
            raise KeyError(f"Unknown enemy pool '{pool_name}'")
        self._current_pool = pool_name

    def add_enemy(self, pool_name: str, template: dict) -> None:
        # Append a new template to a pool when tweaking encounters at runtime.
        self._pools.setdefault(pool_name, []).append(dict(template))

    def next_enemy(self) -> Enemy:
        candidates = self._pools.get(self._current_pool)
        if not candidates:
            raise ValueError(f"Enemy pool '{self._current_pool}' is empty")
        template = dict(self._rng.choice(candidates))
        if "enemy_id" in template:
            enemy_id = template.pop("enemy_id")
            try:
                base = dict(self._enemy_definitions[enemy_id])
            except KeyError as exc:
                raise KeyError(
                    f"Encounter references unknown enemy id '{enemy_id}'"
                ) from exc
            merged = {**base, **template}
        else:
            merged = template
        xp_reward = merged.pop("xp_reward", 0)
        munny_reward = merged.pop(
            "munny_reward",
            merged.pop("gold_reward", 0),
        )
        drops = merged.pop("drops", None)
        level = int(merged.pop("level", merged.pop("enemy_level", 1)) or 1)
        enemy = Enemy(
            **merged,
            xp_reward=xp_reward,
            munny_reward=munny_reward,
            drops=drops,
            level=level,
        )
        return enemy


DEFAULT_ENCOUNTER_POOLS = {
    # Beachfront Heartless patrols for the Destiny Islands starting scenario.
    "destiny_islands_beach": [
        {"enemy_id": "shadow", "level": 1},
        {"enemy_id": "soldier", "level": 2},
    ],
    "destiny_islands_cove": [
        {"enemy_id": "shadow", "level": 2, "xp_reward": 16, "munny_reward": 10},
        {
            "enemy_id": "soldier",
            "level": 3,
            "xp_reward": 32,
            "munny_reward": 18,
            "drops": [
                {
                    "item_id": "champion_belt",
                    "chance": 0.4,
                },
                {
                    "material_id": "mythril_fragment",
                    "chance": 0.65,
                    "amount": 1,
                },
            ],
        },
    ],
}

File: web/src/domain/encounters.ts
import { Enemy } from "./entities";

export type EnemyDrop = {
  itemId?: string;
  materialId?: string;
  chance: number;
  amount?: number;
};

export type EnemyDefinition = {
  name: string;
  hp: number;
  atk: number;
  defense: number;
  cd: number;
  portraitPath: string;
  xpReward: number;
  munnyReward: number;
  drops: EnemyDrop[];
};

export type EncounterTemplate = {
  enemyId?: string;
  level?: number;
  enemyLevel?: number;
  name?: string;
  hp?: number;
  atk?: number;
  defense?: number;
  cd?: number;
  portraitPath?: string;
  xpReward?: number;
  munnyReward?: number;
  goldReward?: number;
  drops?: EnemyDrop[];
};

export const ENEMY_DEFINITIONS: Record<string, EnemyDefinition> = {
  shadow: {
    name: "Shadow",
    hp: 22,
    atk: 2,
    defense: 1,
    cd: 2.5,
    portraitPath: "assets/portraits/enemies/shadow.png",
    xpReward: 12,
    munnyReward: 6,
    drops: [
      { materialId: "bright_shard", chance: 0.5, amount: 1 },
      { materialId: "dark_shard", chance: 0.4, amount: 1 }
    ]
  },
  soldier: {
    name: "Soldier",
    hp: 24,
    atk: 2,
    defense: 1,
    cd: 2.0,
    portraitPath: "assets/portraits/enemies/soldier.png",
    xpReward: 24,
    munnyReward: 12,
    drops: [
      { itemId: "champion_belt", chance: 0.25 },
      { materialId: "mythril_fragment", chance: 0.5, amount: 1 }
    ]
  }
};

export class EncounterPool {
  private pools: Record<string, EncounterTemplate[]> = {};
  private currentPool: string;
  private rng: () => number;
  private enemyDefinitions: Record<string, EnemyDefinition>;

  constructor(params: {
    pools: Record<string, EncounterTemplate[]>;
    defaultPool: string;
    enemyDefinitions?: Record<string, EnemyDefinition>;
    rng?: () => number;
  }) {
    const { pools, defaultPool, enemyDefinitions, rng } = params;
    if (!(defaultPool in pools)) {
      throw new Error(`Unknown enemy pool '${defaultPool}'`);
    }
    Object.entries(pools).forEach(([name, entries]) => {
      this.pools[name] = entries.map((entry) => ({ ...entry }));
    });
    this.currentPool = defaultPool;
    this.rng = rng ?? Math.random;
    this.enemyDefinitions = {};
    const defs = enemyDefinitions ?? ENEMY_DEFINITIONS;
    Object.entries(defs).forEach(([key, value]) => {
      this.enemyDefinitions[key] = { ...value, drops: [...value.drops] };
    });
  }

  getCurrentPool(): string {
    return this.currentPool;
  }

  setPool(poolName: string): void {
    if (!(poolName in this.pools)) {
      throw new Error(`Unknown enemy pool '${poolName}'`);
    }
    this.currentPool = poolName;
  }

  addEnemy(poolName: string, template: EncounterTemplate): void {
    this.pools[poolName] = this.pools[poolName] ?? [];
    this.pools[poolName].push({ ...template });
  }

  nextEnemy(): Enemy {
    const candidates = this.pools[this.currentPool];
    if (!candidates || candidates.length === 0) {
      throw new Error(`Enemy pool '${this.currentPool}' is empty`);
    }
    const template = { ...candidates[Math.floor(this.rng() * candidates.length)] };
    let merged: EncounterTemplate & Partial<EnemyDefinition>;
    if (template.enemyId) {
      const base = this.enemyDefinitions[template.enemyId];
      if (!base) throw new Error(`Encounter references unknown enemy id '${template.enemyId}'`);
      merged = { ...base, ...template };
    } else {
      merged = { ...template };
    }

    const xpReward = merged.xpReward ?? 0;
    const munnyReward = merged.munnyReward ?? merged.goldReward ?? 0;
    const drops = merged.drops ?? null;
    const level = Math.trunc(merged.level ?? merged.enemyLevel ?? 1) || 1;

    const enemy = new Enemy({
      name: merged.name,
      hp: merged.hp,
      atk: merged.atk,
      defense: merged.defense,
      cd: merged.cd,
      portraitPath: merged.portraitPath,
      xpReward,
      munnyReward,
      drops: drops ?? undefined,
      level
    });
    return enemy;
  }
}

export const DEFAULT_ENCOUNTER_POOLS: Record<string, EncounterTemplate[]> = {
  destiny_islands_beach: [
    { enemyId: "shadow", level: 1 },
    { enemyId: "soldier", level: 2 }
  ],
  destiny_islands_cove: [
    { enemyId: "shadow", level: 2, xpReward: 16, munnyReward: 10 },
    {
      enemyId: "soldier",
      level: 3,
      xpReward: 32,
      munnyReward: 18,
      drops: [
        { itemId: "champion_belt", chance: 0.4 },
        { materialId: "mythril_fragment", chance: 0.65, amount: 1 }
      ]
    }
  ],
  traverse_town_first_district: [
    { enemyId: "shadow", level: 3, xpReward: 22, munnyReward: 14 },
    {
      enemyId: "soldier",
      level: 4,
      xpReward: 36,
      munnyReward: 22,
      drops: [{ materialId: "mythril_fragment", chance: 0.6, amount: 1 }]
    }
  ],
  traverse_town_second_district: [
    { enemyId: "soldier", level: 5, xpReward: 44, munnyReward: 28 },
    {
      enemyId: "soldier",
      level: 6,
      xpReward: 48,
      munnyReward: 32,
      drops: [{ itemId: "champion_belt", chance: 0.35 }]
    }
  ]
};
